scripts/fn_paradrop.sqf

```diff
@@ -0,0 +1,198 @@
+private _xDiff = _dropLength/2 * sin _theta;
+private _yDiff = _dropLength/2 * cos _theta;
+// set all coordinates based on DZ points in xy-plane; z coord does not matter
+// double lengths on second DZ coordinate to ensure straight line flight
+private _dropZoneLocs = [[(_dropTarget select 0) + 2*_xDiff,(_dropTarget select 1) + 2*_yDiff,_dropHeight],
+    [(_dropTarget select 1) - _xDiff,(_dropTarget select 1) - _yDiff,_dropHeight]];
+
+//expand waypoint 
+private _dp = _pilotGroup addWaypoint [_dropZoneLocs select 1,-1,currentWaypoint _pilotGroup + 1,""];
+if(!_disableWP2) then {
+    private _de = _pilotGroup addWaypoint [_dropZoneLocs select 0,-1,currentWaypoint _pilotGroup + 2,""];
+};
+
+// Add the waypoint paradrop operation.  Eject automatically gives people a parachute 
+// Adds a key,value pair to the vehicle characteristics hash that can be referenced globally
+// No global variable names required
+_vehicle setVariable ['tac_dropGroups',_dropGroups];
+_vehicle setVariable ['tac_chuteHeight',_chuteHeight];
+
+// note - moveOut instantly pushes them out of vehicle without playing animation
+// this emulates a static line drop dropping very quickly, but needs to be spaced
+// it also means that you need to spawn in a parachute otherwise they freefall to their deaths
+// currently this is set for a drop interval of 1/2 second, allowing AI to keep cohesion
+// but not collide into each other
+
+// if  _groupAjudst is not used, then each group being parachuted will getOut a unit
+// at the same time as the forEach executes in parallel. 
+// Nominally, could use count units _x with _forEachIndex but will not work with groups
+// of different sizes
+
+//  Other notes:
+//  Vehicles are unassigned and allowGetIn is set to false to prevent the paratroopers from
+//  running back to a landed plane or helicopter to get back in again.
+//  Velocity adjustment occurs to differentiate helicopters from planes to alter jump intervals
+
+private _fnc_dropParas = {
+    private _vehicle = vehicle this; 
+    private _paraGroups = _vehicle getVariable ['tac_dropGroups', []];
+    private _chuteHeight = _vehicle getVariable ['tac_chuteHeight',[]];
+    private _dropCount = _paraGroups apply {count units _x};
+    private _groupAdjust = [];
+    _groupAdjust resize (count _dropCount);
+
+    for "_i" from 0 to (count _dropCount)-1 do {
+        private _sum = 0;
+        for "_j" from 0 to _i do {
+            _sum = _sum + (_dropCount select _j);
+        };
+        _groupAdjust set [_i,_sum];
+    };
+
+    private _jumpFreq = 0.3;
+    if(speed _vehicle < 250) then {_jumpFreq = 0.5;};
+
+    {
+        units _x allowGetIn false;
+        private _groupCount = _forEachIndex;
+        {
+            private _jumpDelay = _forEachIndex*_jumpFreq+(_groupAdjust select _groupCount)*_jumpFreq;
+            [{
+                private _man = _this select 0;
+                private _chuteHeight = _this select 1;
```

Comment:
I don't understand the use of `params` in this context.  AFAICT the variables `_man`, `_chuteHeight` and `_vehicle` are not already defined inside of this eventHandler, unless `CBA_fnc_waitAndExecute` also transmits their original names instead of just an array of values.

Reply:
> When used without argument, as shown in main syntax, internal variable `_this`, which is usually available inside functions and event handlers, is used as argument.

Meaning `params [...] == _this params [...]`.

Function arguments are always in `_this`, same for `CBA_fnc_waitAndExecute` arguments.

Reply:
OK - and to be sure I understand this, `CBA_fnc_waitUntilAndExecute`'s blocks are in the same scope as the `CBA_fnc_waitAndExecute` call?  Because what really confused me was using `params` _inside_ the script block you're passing to `CBA_fnc_waitUntilAndExecute`.

Reply:
No, it's a different context. And each block in `CBA_fnc_waitUntilAndExecute` (condition and statement) is different.

`_this` is passed to `CBA_fnc_waitUntilAndExecute`, so it passes all arguments from current scope to it, where I use `params` again to split it out.

Reply:
Got it.